doc/statemachines.tex

```diff
@@ -0,0 +1,310 @@
+  RollForward        & $h$,$p$       & CanAwait    & Idle \\ \hline
+  RollForward        & $h$,$p$       & MustReply   & Idle \\ \hline
+  RollBackward       & $p$,$p$       & CanAwait    & Idle \\ \hline
+  RollBackward       & $p$,$p$       & MustReply   & Idle \\ \hline  
+  FindIntersect      & $points$      & Idle        & Intersect \\ \hline
+  IntersectImproved  & $p_1$,$p_2$   & Intersect   & Idle \\ \hline
+  IntersectUnchanged & $p$           & Intersect   & Idle \\ \hline
+  Done               &               & Idle        & Done \\ \hline
+\end{tabular}
+
+
+\begin{tikzpicture}[->,>=stealth',shorten >=1pt,auto,node distance=4cm,
+                    semithick]
+  \tikzstyle{every state}=[fill=red,draw=none,text=white]
+  \node[state, green, initial]                                   (Idle)      {Idle};
+  \node[state, right of=Idle]                             (Done)      {Done};
+  \node[state, blue, below left of=Idle]                        (CanAwait)  {CanAwait};
+  \node[state, blue, right of=CanAwait]                         (MustReply) {MustReply};
+  \node[state, blue, above of=Idle]                        (Intersect) {Intersect};
+
+%  MsgRequestNext :: ChainSyncMessage header point StIdle (StNext StCanAwait)
+  \draw (Idle)         edge[left, bend right]      node{RequestNext}           (CanAwait);
+  
+%  MsgAwaitReply :: ChainSyncMessage header point (StNext StCanAwait) (StNext StMustReply)
+  \draw (CanAwait)     edge[above, bend right]     node{AwaitReply}            (MustReply);
+  
+%  MsgRollForward :: header -> point -> ChainSyncMessage header point (StNext any) StIdle
+  \draw (CanAwait)     edge[above,bend right=45]     node{RollForward}           (Idle);
+  \draw (MustReply)    edge[right,bend right=45]     node{RollForward}           (Idle);  
+
+%  MsgRollBackward :: point  -> point -> ChainSyncMessage header point (StNext any) StIdle
+  \draw (CanAwait)     edge[above,bend right=80]     node{RollBackward}          (Idle);
+  \draw (MustReply)    edge[right,bend right=80]     node{RollBackward}          (Idle);
+  
+%  MsgFindIntersect :: [point] -> ChainSyncMessage header point StIdle StIntersect
+  \draw (Idle)         edge[right, bend right]    node{FindIntersect}         (Intersect);
+
+%  MsgIntersectImproved  :: point -> point -> ChainSyncMessage header point StIntersect StIdle
+  \draw (Intersect)    edge[above, bend right=45]    node[below = 4mm]{IntersectImproved}     (Idle);
+  
+%  MsgIntersectUnchanged :: point -> ChainSyncMessage header point StIntersect StIdle
+  \draw (Intersect)    edge[above, bend right=80]    node[above = 4mm]{IntersectUnchanged}    (Idle);
+  
+%  MsgDone :: ChainSyncMessage header point StIdle StDone
+  \draw (Idle)         edge[above]                node{Done}                  (Done);
+  
+\end{tikzpicture}
+\\
+
+Client states are green and server states blue.
+Note: Arrows denote state transitions and \emph{not} message delivery.
```

Comment:
I'd be confused by this, since arrows denotes state transitions and messages at the same time.  The state transition occurs upon send (for the sender) / receive (for the receiver), which are apart in time (though not in the logical clock progression, since messages are the light cones of Einstein's special relativity).